### ISSUE TITLE
CNN benchmarks?

### ISSUE THREAD

**Issue by Thanh-Binh** [nyanp/tiny-cnn#75]
Does anyone have a benchmarks results between tiny-cnn and other SWs like Caffe, Theano, cuDNN etc.
for example for:
1. small networks (where I hope tiny-CNN should be better than others)
2. big networks like AlexNet, GoogNet etc, where I hope, other SWs will overperform tiny-cnn?

thanks.

**Comment by edgarriba** [comment]
@Thanh-Binh Please check https://github.com/nyanp/tiny-cnn/issues/197
It's a midterm goal add tiny-cnn to the new benchmarks.

@nyanp Closing this.

**Comment by nyanp** [comment]
@edgarriba 
It's related to #197, but it haven't been solved yet. I'll close it after the new benchmarks added :)

**Comment by bhack** [comment]
@gfursin Do you think that it could be possibile to also collecttive benchmarking tiny-dnn with CK? How we could adopt it?

**Comment by gfursin** [comment]
Hi all. Yes, I will be interested to add Tiny DNN to our crowd-benchmarking and crowd-tuning framework. Actually, I am now heading off to San Francisco where we will be presenting related approach: ARM's workload automation powered by CK: http://schedule.armtechcon.com/session/know-your-workloads-design-more-efficient-systems , but as soon as I am back in a week, I can help with that.

The idea is to add Tiny DNN to CK benchmarking workflow similar to what we already did for Caffe and other applications (TensorFlow in process): http://github.com/dividiti/ck-caffe . This will allow anyone to install and run TinyDNN just with a couple of clicks across diverse hardware, and then start sharing results (see CK wiki for more details about our portable workflows: http://github.com/ctuning/ck/wiki) . 

However, the major challenge is to continue unifying experiments (measured characteristics, design and optimization choices, program and data set properties) - we are gradually working on it as a community by analyzing various realistic workloads (that's why I am interested in TinyDNN)...

I prepared a dummy CK repo for Tiny DNN here: http://github.com/ctuning/ck-tiny-dnn and will add it to our workflow whenever I am back. Will keep in touch!

**Comment by bhack** [comment]
Great I will be also in San Fransico from Wednesday. One particularity of tiny-dnn it is that uses different switchable upstream kernels implementation backends and some of that are tunable at runtime.

**Comment by bhack** [comment]
@gfursin What do you think about metrics, dataset etc. of https://github.com/DeepMark/deepmark and https://github.com/baidu-research/DeepBench. Are you monitoring this two initiatives?

**Comment by gfursin** [comment]
Hi again, @bhack - are you going to ARM TechCon'16 too? Let's have a face-to-face chat there then! I will be very interested to talk to you about all this work... BTW, it's not very easy to figure out your name from your cool alias so just send me an email (grigori.fursin@cTuning.org) and we can arrange a meeting.

**Comment by gfursin** [comment]
By the way, I prepared a simple workflow for Tiny DNN during the flight, and commited it to http://github.com/ctuning/ck-tiny-dnn . Basically, if you have CK installed, you should be able to compile and run a simple bench from this lib via (I wrote it in the readme of this repo):
$ ck pull repo:ck-tiny-dnn
$ ck compile program:tiny-dnn-bench --speed
$ ck run program:tiny-dnn-bench

**Comment by gfursin** [comment]
I know about DeepMark and DeepBench, but, to be honest, didn't have time to look at them deeply - it's in my ToDo list. All my current focus was to make CK robust enough to handle various workloads and benchmarks in a unified way across all existing platforms (Windows, Linux, MacOS, Android) with JSON meta and JSON API which has 5 magic keys ;) to be able to do multi-objective crowd-benchmarking: 'choices', 'characteristics', 'features', 'state', 'deps'. At the same time, we have been improving CK web services to be able to exchange various collected info. I have to admit that this cross-platform support nearly killed me, but it seems to be working fine now. So, after finishing several current urgent projects in November, we will use all our collected experience to get back to universal crowd-benchmarking of any workload (it includes comparison of different implementations of related workloads such as BLAS functions)... Anyway, let's catch up at ARM TechCon - it will be easier to sync on all that ;) ...

**Comment by bhack** [comment]
@gfursin Ok I wrote you from my business account. If you have any spare time try to take a quick look to DeepMark background topic at https://github.com/soumith/convnet-benchmarks/issues/101

**Comment by bhack** [comment]
We got a very interesting face to face meeting with @gfursin at San Fransisco. Ck roadmap it is very interesting. We can start a roadmap to be on par with ck-caffe. How we could go on?

**Comment by gfursin** [comment]
Thanks @bhack - yeap, I really enjoyed our chat in San Francisco! I suggest to have regular meetings in cool places from now on ;) ... 

Jokes apart, we continue working on the following in the CK-Caffe (that should be later applicable to other frameworks): I am now adding a new API to the CK program workflow to be able to collect optimization choices (OpenCL/CUDA/OpenMP) at finer-grain level together with profiling information (network layers/BLAS functions) and data set "features" (related to performance optimization not recognition) such as image/matrix sizes. 

At the moment, we use a small instrumentation Run-Time library xOpenME that I have developed and integrated with the CK especially for such kind of purposes. It's not optimized at all and may have some overheads, but it is stable, works across Windows, Linux, MacOS and Android, and more importantly, it dumps JSON file in the CK format, so it's easy to inject ouput to the CK benchmarking and autotuning workflow. If you have CK installed, you can find it via:
$ ck pull repo:ck-autotuning
$ ck find package:lib-rtl-xopenme

You can find how it is used in the susan corners program (image corners/edges detection program) that I used a lot in my past research on auto-tuning (you will see how we extract image features there):
$ ck pull repo:ctuning-programs
$ ck find program:cbench-automotive-susan
and search for 'openme' in susan.c

We also integrated simple open-source OpenCL profiles from my colleague @psyhtest to be able to transparently capture behavior of OpenCL-based BLAS and Caffe (see https://github.com/dividiti/dvdt-prof - it is also available as a package in the CK).

As for optimizations, at the moment, we connect our CK autotuner via #defines and environment variables, so one will need to expose available optimizations (such as LWS_X, LWS_Y, padding and other parameters in OpenCL) via our worklow.

Finally, just like anyone else, I started working on a proper validation of program output (not only exact match, but also using thresholds for floating point results).

My timeline is to finish all that by the end of this year focusing on Caffe (we are now awfully busy finishing couple of very important projects and they will be open-sourced too). However, the most important part is the universal CK methodology here which should allow to add any other workload to CK reusing or slightly altering existing interfaces. So if we will be able to do it for Caffe, we can do it for any other framework too.

However, for you not to wait until we prepare and stabilize new interfaces, you can already try to instrument TinyDNN using OpenME framework (or using any other solution since we use postprocessing python script after workload execution to convert raw results to CK format) to expose required characteristics per network layer, or expose related data sets, etc. We can also provide an extra module in the CK which will try to unify naming of NN layers across different implementation to know that their algorithmic behavior is equivalent (otherwise everyone is using their own terms)... 

Sorry if something is not yet clear - I just tried to quickly write a few ideas here. We are now in the middle of unifying CK-Caffe interfaces and benchmarking, which is quite tough. That's why, I now want to stabilize our current solution before engaging with open discussions to avoid changing APIs in the middle of our project (we have a tough deadline) ;) ... But I will keep in touch as soon as we have more progress on APIs since I do realize now that our project can also be useful for (https://github.com/soumith/convnet-benchmarks/issues/101) ...

Will keep in touch!

**Comment by bhack** [comment]
Thank you for this big recap. I'm still in Sisco for some days I will try to recap all the pices when I'will come back next week.
About standardizzation an effort was started between hadware vendors to try sandardize [Dnn at Kronos](http://www.khronos.org/registry/vx/extensions/neural_network/html/index.html). This is positive as initiative  but one of the point against it is that probably will be always far or less behind state of the art so I don't know form a tuning benchmarkig perspective how this could be interesting.

**Comment by gfursin** [comment]
Sure, sounds good! 
Yes, I saw the announcement about Kronos' initiative to standardize DNN - sounds interesting, so will be monitoring this activity. 
By the way, getting back to DNN benchmarking - you may also be interested to have a look at https://github.com/dividiti/gemmbench - it is also a part of our initiative to unify GEMM crowd-tuning via CK. I am now using all this practical experience to develop a new CK-based universal workload crowd-benchmarking and crowd-tuning including DNNs ...
Have a safe trip home!

**Comment by bhack** [comment]
The problem with gemm is that [batched operations](http://www.netlib.org/utk/people/JackDongarra/WEB-PAGES/Batched-BLAS-2016/Day1/01_Tomov_Overview.pdf) ibterfaces are still not standardized and are not available in every libs.
Also direct convoutions are still more relevant in Dnn that gemm actually.

**Comment by gfursin** [comment]
Yes, you are right... Let's start from what is already available and relatively stable...

**Comment by Randl** [comment]
https://github.com/tiny-dnn/tiny-dnn/issues/277 is related too
